### PR TITLE
Fix S1313 FN: Support utf-8 literals and add support utf8 string literals in CSharpFacade

### LINKED ISSUE #6376: Fix S1313 FN: Support utf-8 literals
### Description

FN for S1313 (HardcodedIpAddress).

Using C#11 utf-8 string literals, the rule is not raising an issue.

### Repro steps

```csharp
            var ip2 = "192.168.0.1"u8; // FN
            var ip3 = """192.168.0.1"""u8; // FN
            var ip4 = """
                192.168.0.1
                """u8; // FN

```

### Expected behavior

The rule should raise an issue.

### Actual behavior

The rule does not raise an issue.

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
Fixes #6376  (this was a byproduct of adding the utf8 string literals to the CSharp Facade)

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs:57]: I am not sure this is a 100% correct, as UTF8 Strings are strings and at the same time they are not. I personally would not add the `Utf8StringLiteralExpression` here but rather handle them separately.

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs:57]: From some tests I did - from an AST point of view, they seem to have similar behavior (almost the same).
Take a look at this [example in SharpLab](https://sharplab.io/#v2:C4LghgzsA0AmIGoA+ABATARgLAChcoBYACAJTAHcBlYAJwEsA7AcwBk7gBTGsAGwgAoUGAAxEArgwDWDAPbkGABTA0YRIaOlzFy4AEpcAb1xETakUToAHDEQC8RAEROMATjQA6DADYAHO+GeTg4A3MamAG7KFpYAzHaOADoAHjEYyTEu6WjJaACi6WkpXuk+OfkpwmUFDmI+oTgAvkA=).

In the semantic universe, they are not the same - but I did not see any suspicious case that can get us bitten in the future so far.

Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs:57]: They might have similar behaviour and yet they are completely different things. And when I use `StringLiteralExpression` from the syntax kind facade I expect really to only handle string literals. There are rules and cases when we only want to handle one of them and not both. One of those examples could be the rule which says not to call `ToString` on a string literal as it is redundant. While it is redundant for the `StringLiteral` it is not for the `UTF-8 string literal`. And basically if one would want to implement that rule using the `SyntaxKindFacade` it would not be possible. I either suggest to either revert the Facade to its previous state and handle the SyntaxKind separately or Add a new property to the facade which clearly states that it contains both "StringLiterals". WDYT?

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs:57]: I don't agree with Csaba here. We already face the same problem with interpolated strings, which can be of type string or FormattableString depending on the target type:
```cs
FormattableString fs = $""; 
string st = $""; 
```
We don't distinguish here also.

If you see a string literal, you can not be sure it is assigned to a `string` either:
```cs
object o = "";
IComparable c = "";
T t = "";

public class T
{
    public static implicit operator T(string v) => new T();
}
```
So you always have to check the target type and the only difference with `u8` is that the targets are now
for string literals:

* string
* object
* IComparable
* IEnumerable
* Any type with an implicit conversion from string
 
and for ut-8 literals:

* `ReadonlySpan<byte>`

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs:57]: > One of those examples could be the rule which says not to call ToString on a string literal as it is redundant.

The rule would be to not call `ToString` on a string **type** not on a string **literal** (and besides that, there is a ToString overload on string which is sensible to call).

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs:75]: Starting with .NET 4.7.1 there is [Enumerable.Append<TSource>(IEnumerable<TSource>, TSource) Method (System.Linq)](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.append?view=netframework-4.8). Can you use that instead or do we target lower versions of the framework?
```suggestion
            var kinds = Language.SyntaxKind.StringLiteralExpressions.Append(Language.SyntaxKind.InterpolatedStringExpression).ToArray();
```

**Review by martin-strecker-sonarsource — COMMENTED** [review]
Looks good from my side. Waiting for a response from @csaba-sagi-sonarsource regarding his concerns. One improvement might be possible but I'm not sure.

File: analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs:57]: @mary-georgiou-sonarsource You should have a look at where StringLiteralExpressions is used and add tests for ut8 literals in such rules. We can decide from there if @csaba-sagi-sonarsource concerns are valid today already.

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs:57]: It seems that two Hotspots register for `StringLiteralExpressions`. 
I added UTs for utf-8 string literals to both.

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs:57]: In the CSharp rules additionally, **S4784** `UsingRegularExpressions`  is using the "GetStringValue" - to check strings given as input to Regex Object creation or invocation.
I see no issue there regarding the facade update.

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs:75]: I see we are targetting `net46;netstandard2.0`.

File: analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs:57]: Fine by me.

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs:57]: > It seems that two Hotspots register for StringLiteralExpressions coming from the Facade.
I added UTs for utf-8 string literals to both.

I only see PublicWritableDirectories. What is the other one or do I miss something?

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs:75]: That's unfortunate.

File: analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Facade/CSharpSyntaxKindFacade.cs:57]: Never mind. Its HardcodedIpAddress.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/PubliclyWritableDirectories.CSharp11.cs
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/PubliclyWritableDirectories.CSharp11.cs:21]:
```suggestion
        void Utf8StringLiterals()
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/PubliclyWritableDirectories.CSharp11.cs
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/PubliclyWritableDirectories.CSharp11.cs:25]: To be in sync with the base test file or just keep one usage and remove the othersbut keep one `@""u8`
```suggestion
            tmp = "%TEMP%\\f"u8;                // Noncompliant
            tmp = "/tmp/"u8;                    // Noncompliant
            tmp = "/tmp"u8;                     // Noncompliant
            tmp = "/var/tmp/f"u8;               // Noncompliant
            tmp = "/usr/tmp/f"u8;               // Noncompliant
            tmp = "/dev/shm/f"u8;               // Noncompliant
            tmp = "/dev/mqueue/f"u8;            // Noncompliant
            tmp = "/run/lock/f"u8;              // Noncompliant
            tmp = "/var/run/lock/f"u8;          // Noncompliant
            tmp = "/Library/Caches/f"u8;        // Noncompliant
            tmp = "/Users/Shared/f"u8;          // Noncompliant
            tmp = "/private/tmp/f"u8;           // Noncompliant
            tmp = "/private/var/tmp/"u8;        // Noncompliant
            tmp = "C:\\Windows\\Temp\\f"u8;     // Noncompliant
            tmp = "C:\\Temp\\f"u8;              // Noncompliant
            tmp = "C:\\TEMP\\f"u8;              // Noncompliant
            tmp = "C:\\TMP\\f"u8;               // Noncompliant
            tmp = @"/tmp/f"u8;                  // Noncompliant
            tmp = "D:\\Windows\\Temp\\f"u8;     // Noncompliant
            tmp = "\\\\Server_Name\\Temp\\f"u8; // Noncompliant
            tmp = "\\Windows\\Temp\\f"u8;       // Noncompliant
            tmp = @"C:\Windows\Temp\f"u8;       // Noncompliant
            tmp = @"D:\Windows\Temp\f"u8;       // Noncompliant
            tmp = @"\\Windows\Temp\f"u8;        // Noncompliant
            tmp = @"\Windows\Temp\f"u8;         // Noncompliant            
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**
